### PR TITLE
TreeViewList Component

### PR DESCRIPTION
**Pull request by syedsalehinipg**
Contributes to [TD-2030](https://sce.myjetbrains.com/youtrack/issue/TD-2030) Migrating VIRTO.DATA to Remix

## Changes

The PR introduces a new TreeViewList component that renders a tree view list. This component can be used to display hierarchical data in a tree structure.

- The TreeViewList component accepts several props including items, selected, searchTerm, defaultExpanded, and onSelectionChange. These props allow for customization of the tree view list.
- The TreeViewList component supports tooltips. When a tree item is hovered over, a tooltip with the item's name and additional information is displayed.
- The TreeViewList component includes a search function that can filter the items in the tree view list based on a search term.

## UI/UX

image

image

image

## JSON Structure

The JSON structure represents a list of items, where each item can have children and options. Here's an example:

```json
[
  {
    "name": "BRK",
    "id": "64f1b6fd511d08b5f6bc2a1e",
    "children": [
      {
        "name": "MCbooster",
        "children": [
          {
            "name": "HydESPModel",
            "children": [
              {
                "name": "PumpMaxDelivery",
                "options": [],
                "tooltip": "Maximum delivery of the two hydraulic pumps..."
              }
            ]
          }
        ]
      },
      {
        "name": "Pedal",
        "children": [
          {
            "name": "Ratio",
            "options": [],
            "tooltip": "The pedal ratio amplifies the force of the brake pedal..."
          },
          {
            "name": "ResponseTime",
            "options": [],
            "tooltip": "Period of time from pressing the pedal till the brake pressure begins to build up."
          }
        ]
      }
    ]
  },
  {
    "name": "ELE",
    "children": [
      {
        "name": "Battery",
        "children": [
          {
            "name": "HV",
            "children": [
              {
                "name": "Capacity",
                "options": [],
                "tooltip": "This is the total capacity of the board high voltage battery"
              }
            ]
          }
        ]
      }
    ]
  },
  {
    "name": "TRM",
    "children": [
      {
        "name": "Efficiency1D",
        "options": []
      },
      {
        "name": "GearSpred",
        "options": []
      }
    ]
  }
]
```

## Author checklist before assigning a reviewer

- [x] Reviewed my own code-diff.
- [ ] Branch has been run in docker.
- [x] PR assigned to me or an appropriate delegate.
- [x] Relevant labels added to the PR.
- [x] Appropriate tests have been added.
- [x] Lint and test workflows pass.

### CODE REVIEW

**Comment by mattcorner** [comment]
We also have a TreeView in VIRTO.BUILD. Can we check that this react-ui component incorporates all of the uses cases from BUILD as well so that we can switch to this common component in the future. @evoinea any unique use cases that you are aware of?

image

**Comment by m4manjesh** [comment]
@syedsalehinipg  Did a bash test and the UI looks good and works fine. As I am away for a couple of days if you are adding more details based on the above comment made by Matt into this PR please ask someone to do the final review.

File: src/TreeViewList/TreeViewList.tsx [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+ *
+ * @template T The type of the options array elements.
+ * @param data - The flat list of items to transform into a tree.
+ * @returns The tree structure built from the input items.
+ * @example buildTree(items) // returns TreeNode[]
+ */
+function buildTree<T>(data: Item<T>[]): TreeNode[] {
+  const itemsTree: TreeNode[] = [];
+
+  // Iterate through each item
```
Comment (evoinea) [inline src/TreeViewList/TreeViewList.tsx:203]:
```suggestion
  // iterate through each item
```

File: src/TreeViewList/TreeViewList.tsx [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+  data.forEach(item => {
+    // create node for this root model
+    const name = item.name ? item.name.toString() : "default";
+    const tooltip = item.tooltip ? item.tooltip.toString() : undefined;
+    const parentNode = createNode(name, name, tooltip);
+    itemsTree.push(parentNode);
+
+    if (Array.isArray(item.options)) {
```
Comment (evoinea) [inline src/TreeViewList/TreeViewList.tsx:211]:
```suggestion
// parse options and children recursively
    if (Array.isArray(item.options)) {
```

File: src/TreeViewList/TreeViewList.tsx [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+            borderLeft: `1px solid ${alpha(theme.palette.text.primary, 0.1)}`,
+            color: theme.palette.text.primary,
+            padding: "5px"
+          })}
+        />
+      </Tooltip>
+    );
+  };
+
+  // Recursive search function
```
Comment (evoinea) [inline src/TreeViewList/TreeViewList.tsx:83]:
```suggestion
  // recursive search function
```

File: src/TreeViewList/TreeViewList.tsx [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+
+    const searchInItem = (item: Item<T>): Item<T> | null => {
+      const match = terms.some(term =>
+        typeof item.name === "string"
+          ? item.name.toUpperCase().includes(term)
+          : false
+      );
+
+      if (match) {
+        // If the item matches any of the search terms, return the whole item
```
Comment (evoinea) [inline src/TreeViewList/TreeViewList.tsx:98]:
```suggestion
        // if the item matches any of the search terms, return the whole item
```

**Review by evoinea — CHANGES_REQUESTED** [review]
Looks good. Minor formatting suggestions to address.

**Comment by evoinea** [comment]
Have you demoed this new component to @Sowbhagya-ipg?

**Review by evoinea — APPROVED** [review]
LGTM

**Pull request merged**